command/agent/cache/lease_cache.go

```diff
@@ -0,0 +1,813 @@
+		Ctx:        baseCtx,
+		CancelFunc: baseCancelFunc,
+	}
+
+	return &LeaseCache{
+		proxier:     conf.Proxier,
+		logger:      conf.Logger,
+		db:          db,
+		baseCtxInfo: baseCtxInfo,
+	}, nil
+}
+
+// Send performs a cache lookup on the incoming request. If it's a cache hit,
+// it will return the cached response, otherwise it will delegate to the
+// underlying Proxier and cache the received response.
+func (c *LeaseCache) Send(ctx context.Context, req *SendRequest) (*SendResponse, error) {
+	// Compute the index ID
+	id, err := computeIndexID(req)
+	if err != nil {
+		c.logger.Error("failed to compute cache key", "error", err)
+		return nil, err
+	}
+
+	// Check if the response for this request is already in the cache
+	index, err := c.db.Get(cachememdb.IndexNameID, id)
+	if err != nil {
+		return nil, err
+	}
+
+	// Cached request is found, deserialize the response and return early
+	if index != nil {
+		c.logger.Debug("returning cached response", "path", req.Request.URL.Path)
+
+		reader := bufio.NewReader(bytes.NewReader(index.Response))
+		resp, err := http.ReadResponse(reader, nil)
+		if err != nil {
+			c.logger.Error("failed to deserialize response", "error", err)
+			return nil, err
+		}
+
+		return &SendResponse{
+			Response: &api.Response{
+				Response: resp,
+			},
+			ResponseBody: index.Response,
+		}, nil
+	}
+
+	c.logger.Debug("forwarding request", "path", req.Request.URL.Path, "method", req.Request.Method)
+
+	// Pass the request down and get a response
+	resp, err := c.proxier.Send(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the namespace from the request header
+	namespace := req.Request.Header.Get(consts.NamespaceHeaderName)
+	// We need to populate an empty value since go-memdb will skip over indexes
+	// that contain empty values.
+	if namespace == "" {
+		namespace = "root/"
+	}
+
+	// Build the index to cache based on the response received
+	index = &cachememdb.Index{
+		ID:          id,
+		Namespace:   namespace,
+		RequestPath: req.Request.URL.Path,
+	}
+
+	secret, err := api.ParseSecret(bytes.NewBuffer(resp.ResponseBody))
+	if err != nil {
+		c.logger.Error("failed to parse response as secret", "error", err)
+		return nil, err
+	}
+
+	isRevocation, err := c.handleRevocationRequest(ctx, req, resp)
+	if err != nil {
+		c.logger.Error("failed to process the response", "error", err)
+		return nil, err
+	}
+
+	// If this is a revocation request, do not go through cache logic.
+	if isRevocation {
+		return resp, nil
+	}
+
+	// Fast path for responses with no secrets
+	if secret == nil {
+		c.logger.Debug("pass-through response; no secret in response", "path", req.Request.URL.Path, "method", req.Request.Method)
+		return resp, nil
+	}
+
+	// Short-circuit if the secret is not renewable
+	tokenRenewable, err := secret.TokenIsRenewable()
+	if err != nil {
+		c.logger.Error("failed to parse renewable param", "error", err)
+		return nil, err
+	}
+	if !secret.Renewable && !tokenRenewable {
```

Comment:
You should still cache this for the lifetime of the secret/token, just not start a renewer process for it.

Reply:
Maybe to put it another way, you may want to just remove this check, then in the cases below, add to index but don't start the renewer if it's not renewable.

Reply:
Currently only the renewer goroutine when returning evicts entries from cache. The reason for designing this way is that we don't need to worry about invalidating the cache. There are no timers anywhere. But if we start caching non-renewable secrets, that won't happen. Although this seems like a very valid scenario to cache, we are simply not.

In fact the kind of items that are being cached (with v1 of this feature) are pretty scoped. Only the tokens that are created through the agent will be cached. This means that all the older tokens that transits through agent will be pass-through requests. Only the leases created by tokens that are already managed by the agent are cached. This gives more control on the items that are cached. However, this will most certainly change going forward and we'll add more and more cacheable scenarios will be added to the cache. Another goal with this approach was to ensure that the new feature will be a low touch thing, as in it won't disrupt the already existing tokens and lease management and is applicable only to new ones.

Reply:
I would assume that revocation also evicts from the cache and stops any renewer activity?

But yeah I think a future direction here should be that anything with a lease gets cached regardless of renewability.

Reply:
Yes, token or lease revocation results in the renewer getting killed, in-turn resulting in the eviction of respective entries from cache.